### DISCUSSION TITLE
What can I do in an emergency if a plugin I need isn't on the Internet anymore, temporarily or permanently?

### DISCUSSION THREAD

**Discussion by SethTisue** [sbt/sbt#7206] [General]
this came up over at #7202. I'm making it its own discussion so we don't clutter up the parent ticket with details and discussion

there are two parts to the problem

1) getting the JAR (and POM and so forth) you need if you don't already have it
2) putting it somewhere sbt can find it

1 might be hard depending on the files you need. maybe you can find someone who has the artifacts in their local cache.

how about 2?

## a couple possible workaround paths

if anyone is really desperate for an immediate workaround, consider the following very hastily written rough notes I don't have time right now to polish:

2 is actually doable afaik, you can just throw it in the right place inside `~/.ivy2` or `~/.m2` and sbt will find it. you could throw up any kind of HTTP server anywhere on the internet that's serving the JARs and POMs and everything out of the right directory structure and add it a resolver. Maven repositories don't have a bunch of magic, they're just HTTP servers serving directory trees

if you need it on CI rather than a local machine, you could upload the JARs to your CI machine, or add CI steps that grab them from wherever URL you put them at, and then from there stash them in `~/.ivy2` and `~/.m2`

all of this is just gritty annoying grunt work requiring some knowhow, there aren't any fundamental obstacles that I can see
if you are desperate and really need something right away

## other possible workaround paths?

I hope people will leave them in the comments below

**Comment by joel-apollo** [comment]
One thing you'll run into if you're trying to move your local ivy cache into an off the shelf maven repo is 
a) ivy doesn't have a pom file, but you can generate it with ivy `makepom`
b) getting the qualifiers for maven and sbt version right is tricky, I couldn't make it work with our commercial repo tool (makepom doesn't add them)

However, you could move your ivy cache to an S3 bucket and let your CI copy it before running it

**Comment by turb** [comment]
If you have the jar file stored somewhere, a workaround is to include it directly in your project (here with sbt-dynver):
```
mkdir project/lib
cp ~/.ivy2/cache/scala_2.12/sbt_1.0/com.dwijnand/sbt-dynver/jars/sbt-dynver-4.1.1.jar project/lib
```

Then remove/comment out `addSbtPlugin("com.dwijnand" % "sbt-dynver" % "4.1.1")` from `project/plugins.sbt`.

**Reply by joel-apollo** [reply]
Great tip! Unfortunately this doesn't seem to work so well if there are plugins hosted elsewhere that rely on dependencies that are in the plugins repo. Specifically the `"com.typesafe.play" % "sbt-plugin"` But maybe I can find the jar elsewhere.
